src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -1063,12 +1069,41 @@ public void EnsureRenderedBodyOrSections()
 
         public void BeginContext(int position, int length, bool isLiteral)
         {
+            const string BeginContextEvent = "Microsoft.AspNet.Mvc.Razor.BeginInstrumentationContext";
+
             PageExecutionContext?.BeginContext(position, length, isLiteral);
+            if (DiagnosticSource != null && DiagnosticSource.IsEnabled(BeginContextEvent))
+            {
+                DiagnosticSource.Write(
+                    BeginContextEvent,
+                    new
+                    {
+                        httpContext = Context,
+                        path = Path,
+                        isPartial = IsPartial,
+                        position = position,
+                        length = length,
+                        isLiteral = isLiteral,
+                    });
+            }
         }
 
         public void EndContext()
         {
+            const string EndContextEvent = "Microsoft.AspNet.Mvc.Razor.EndInstrumentationContext";
+
             PageExecutionContext?.EndContext();
+            if (DiagnosticSource != null && DiagnosticSource.IsEnabled(EndContextEvent))
```

Comment:
Yup, I could do that now that I actually understand how `.?` works for value types - https://github.com/aspnet/HttpAbstractions/pull/461#discussion_r43961310